presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java

```diff
@@ -392,6 +409,40 @@ public void enforceQueryMaxRunTimeLimits()
     }
 
     /**
+     * Enforce max driver number at the query level
+     */
+    public void enforceQueryMaxTasksLimits()
+    {
+        for (QueryExecution query : queries.values()) {
+            if (query.getQueryInfo().getState().isDone()) {
+                continue;
+            }
+            int queryMaxDrivers = SystemSessionProperties.getQueryMaxDrivers(query.getSession());
+            int totalDrivers = query.getQueryInfo().getQueryStats().getTotalDrivers();
+            if (queryMaxDrivers < totalDrivers) {
+                query.fail(new PrestoException(EXCEEDED_DRIVER_NUMBER_LIMIT, "Query exceeded maximum driver number limit of " + queryMaxDrivers));
```

Comment:
The only way to do that is to have the connector enforce the limit. Presto has no knowledge of the number of splits or their size, until the query executes